examples_tests/13.BRDF_Explorer/deriv_map_gen.comp

```diff
@@ -0,0 +1,45 @@
+#version 450 core
+
+layout (local_size_x = 16, local_size_y = 16) in;
+
+layout (binding = 7) uniform sampler2D bumpMapSampler;
+layout (binding = 0, rg8_snorm) uniform image2D derivativeMapImage;
+
+layout (location = 0) uniform float uHeightScaleFactor;
+
+shared float smem[324];//18*18
+
+int getAddr(in ivec2 threadID)
+{
+	return 18*(threadID.y+1) + threadID.x+1;
+}
+
+void main()
+{
+	const ivec2 G_IDX = ivec2(gl_GlobalInvocationID.xy);
+	
+	const ivec2 bumpMapSz = textureSize(bumpMapSampler, 0);
+
+	const ivec2 LC_IDX = ivec2(gl_LocalInvocationID.xy);
+	smem[getAddr(LC_IDX)] = texelFetch(bumpMapSampler, G_IDX, 0).x * uHeightScaleFactor;
+	if (LC_IDX.x == 0 || LC_IDX.x == 15) // TODO how not to use if-statements??
+	{
+		ivec2 offset = ivec2(mix(ivec2(1, 0), ivec2(-1, 0), LC_IDX.x == 0));
+		smem[getAddr(LC_IDX+offset)] = texelFetch(bumpMapSampler, G_IDX+offset, 0).x * uHeightScaleFactor;
+	}
+	if (LC_IDX.y == 0 || LC_IDX.y == 15)
+	{
+		ivec2 offset = ivec2(mix(ivec2(0, 1), ivec2(0, -1), LC_IDX.y == 0));
+		smem[getAddr(LC_IDX+offset)] = texelFetch(bumpMapSampler, G_IDX+offset, 0).x * uHeightScaleFactor;
+	}
+	
+	barrier(); // AFAIK execution barrier implies memory barrier
```

Comment:
does it?

Reply:
I think its best to put the shared memory barrier in, I've read the Vulkan Memory Model Appendix and I think execution synchronisation and availability+visibility of writes are two separate things (although visibility/availability requires execution synch)

Reply:
@Crisspl ?

Reply:
i remember that while you were reading vulkan spec some long time ago you found out that exec barrier implies memory barrier, you tried it out and it was working and saving quite nice % of performance. So i take it it's not true, at lest in opengl?

Reply:
resolved
but give me an answer for the question :)

Reply:
noooo.... I read in the Vulkan spec that exec barrier implies only exec barrier and memory barrier is needed to ensure transactions are visible and available.

I did however read that `coherent` qualified variables do not need `memoryBarrier` because they are well, coherent.

Reply:
https://www.khronos.org/opengl/wiki/Compute_Shader#Shared_memory_coherency

> The usual set of memory barriers is available to compute shaders, but they also have access to memoryBarrierShared(); this barrier is specifically for shared variable ordering. groupMemoryBarrier() acts like memoryBarrier(), ordering memory writes for all kinds of variables, but it only orders read/writes for the current work group.

> While all invocations within a work group are said to execute "in parallel", that doesn't mean that you can assume that all of them are executing in lock-step. If you need to ensure that an invocation has written to some variable so that you can read it, you need to synchronize execution with the invocations, not just issue a memory barrier (you still need the memory barrier though).

Reply:
Basically
`memoryBarrierShared` < `groupMemoryBarrier` < `memoryBarrier`

Also 
> After ensuring ordering, the other element that is needed for visibility is special GLSL syntax. The image or buffer variable being written to and read from must be qualified by the coherent qualifier. Note that both the writer and reader shaders must qualify their variables properly; otherwise, nothing is guaranteed.

> Qualifying the variable with coherent alone is not enough however. You also need to use a memory barrier;

So basically you need both `coherent` and a memory-barrier.

Unless you are doing *atomic* operations *only* on *coherent* qualified variables, then no memory barrier is necessary.